### ISSUE TITLE
Multi-system Dependency Resolution

### ISSUE THREAD

**Issue by YaqiWang** [idaholab/moose#6369]
We met this situation in our simulations where, a postprocessor depends on the evaluation of an aux variable then another aux variable depends on the postprocessor all having the same execute_on. Currently in MOOSE, this cannot be resolved properly because the postprocessor will be always evaluated after aux variable evaluation. Typically this situation can be resolved by changing the execute_on. It would be nice that MOOSE can issue an error on this situation instead of letting users figure out the error in a really hard way. Thanks!

**Comment by YaqiWang** [comment]
The situation is simply auxkernel=>userobject=>auxkernel on the same `execute_on`. Currently we only have such an ordering: `preaux userobject=>auxkernel=>postaux userobject`. Unless we had `postUO auxkernel`, we need the error message. I think this should be a critical issue.

**Comment by permcody** [comment]
While it is a problem, it is currently outside the design goal of MOOSE. We simply don't have arbitrary object dependencies and we aren't set up for that just yet. It would require a unified "coupling" interface so that every object's dependency was at a higher level in the inheritance. Any fix to this will require a huge refactor or some stop gap hack to combine across systems like we are doing now (and it's not pretty). I'm not saying we won't do it, I just don't have a good way of doing this cleanly or generally right now.

**Comment by friedmud** [comment]
@permcody : do you think it would be hard to throw an error in this case? I
honestly don't know if it would be difficult/impossible or not.

**Comment by permcody** [comment]
For a specific combo, no. I'll look into it.

**Comment by YaqiWang** [comment]
@permcody Sorry for iterating, but we really want this urgently. We see several use cases with this error including the sprint problem. Meanwhile, it will be nice to issue an error on the another dependency:
Aux=>pp1=>pp2, where Aux is not directly depending on pp2, so pp2 is in the postaux group.

**Comment by permcody** [comment]
Will you point me to an input file that should fail but does not? I'll look to see if I can make it work for you.

**Comment by YaqiWang** [comment]
Let me create a moose test for these two cases. Will let you know once it is done.

**Comment by YaqiWang** [comment]
@permcody Two tests are added. Please mark do-not-merge in my PR. Thanks.

**Comment by permcody** [comment]
Thanks

**Comment by permcody** [comment]
Renaming this issue. The previous title wasn't useful

**Comment by YaqiWang** [comment]
This seems still an outstanding issue. We do not need a resolution for now but need an error.

**Comment by lindsayad** [comment]
I think this ticket has significant overlap with the newer ticket #14762. I'm going to mark this as a duplicate and close. If others disagree with me, then we can reopen.

**Comment by YaqiWang** [comment]
If adding the error is difficult, I will not insist to have it. But if it is not that difficult, for example, can be done in a week, I suggest we add that first before we pursue #14762. I think this is a good strategy. We could potentially fix some tests by doing this.

**Comment by lindsayad** [comment]
I am basing the issue overlap mostly on the issue titles. I added to the design description of #14762 that we should try to emit intelligent errors when possible until we have arbitrary dependency ordering in place
